alsa-scarlett-gui: fixed desktop entry

inyourfoss opened the pull request:
Fixed desktop entry of the app so that it shows up in the menu. 

###### Description of changes
In place search and replace was implemented in the postInstall phase using sed.

@SebTM as requested the solution now is all in one file.

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

inyourfoss committed:
alsa-scarlett-gui: fixed desktop entry

Fixed desktop entry of the app so that it shows up in the menu.
In place search and replace was implemented in the postInstall phase
using sed.

SebTM commented on pkgs/applications/audio/alsa-scarlett-gui/default.nix:
```diff
@@ -24,6 +24,9 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ pkg-config wrapGAppsHook4 ];
   buildInputs = [ gtk4 alsa-lib ];
+  postInstall = ''
+      sed -Ei "s/(Exec\s?=\s?).*/\1alsa-scarlett-gui\nTryExec=alsa-scarlett-gui/" $out/share/applications/vu.b4.${pname}.desktop
```
From what I remember, can't find the exact comment in one of my MR/or where I participated substituteInPlace is preferred over sed, but will leave it to the merger if it's needed as I don't have merge permissions ✌🏻

gnumber13 replied:
Just to give my 2cents to what speaks for sed over substituteInPlace (as far a I could investigate):

First I think it is not that easily testable. Which you really need for non trivial stuff. sed is a standard unix tool which you can easily test for the correct result on every unix/linux command line.

Secondly you could only use substituteInPlace on the exact match of the entire line (as I could not find any documentation about pattern matching)  which means worst case you would have to change it in the default.nix on every release. sed matches the Line from "Exec=" to the end of the line (not matter what comes after  "Exec=") and replaces that. That should make it more resistant to changes later on.

SuperSandro2000 replied:
sed also does nothing if it matches nothing unlike substituteInPlace which logs a warning.

Anyway, since we are patching multiple lines it would be easier to follow the change, if it would be a patch

SebTM replied:
btw. I don't think try-exec is really needed as we discarded the %F stuff etc.

SuperSandro2000 replied:
Can we remove it then?

gnumber13 replied:
I will test if the tryexec key can be removed and report back.

Just to summarize the most preferred way would be to use substituteInPlace to substitute the line with the Exec key exactly and replace that with the desired line?
And how would I go about applying the changes? A new PR?

SebTM reviewed (approved):
LGTM

gnumber13 commented:
Hey @SebTM ,
somehow the check for aarch64-darwin has been running for 3 days.
Will someone have to intervene?
And when do the merges happen usually? 
Greets

SebTM commented:
@ofborg build alsa-scarlett-gui

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1033

SebTM commented:
> Hey @SebTM , somehow the check for aarch64-darwin has been running for 3 days. Will someone have to intervene? And when do the merges happen usually?

I tried to restart it now.

> 

`but will leave it to the merger if it's needed as I don't have merge permissions ✌🏻`
You can post them also [here](https://discourse.nixos.org/t/prs-already-reviewed/2617/1033) after you got one approval at least (at best from a maintainer), I did that now ✌🏻 

Otherwise you can use an overlay to fix it for for your system before merge: https://nixos.wiki/wiki/Overlays

SuperSandro2000 commented:
> Hey @SebTM , somehow the check for aarch64-darwin has been running for 3 days. Will someone have to intervene? And when do the merges happen usually? Greets

Can be ignored if it never finishes.

inyourfoss committed:
alsa-scarlett-gui: fix for deskop integration

Now pretty much any DE should support the desktop entry file.
The targetpath for the "Exec=" key is now absolute and targets the
binary in the nix store. This also applies to the "Icon=" key.
Also "substitueInPlace" is used to manipulate the default desktop entry
file.

gnumber13 commented:
FYI: @SuperSandro2000 @SebTM 
I added a new commit that manipulates the desktop entry file using "substituteInPlace".

During testing I realized it would be best to have the absolute paths to the executable in the nix store.
Since in my testing KDE Plasma had issues picking up the desktop entry containing the relative path.

I think that might be the more robust solution.

Cinnamon, Gnome and XFCE were working either way.

inyourfoss committed:
alsa-scarlet-gui: light refactoring
Done some light refactoring for better consitence in the code.

SebTM commented:
For the binary I see a need for certain environments to use full paths about the icon thing I'm unsure I found it a few times but in general they are collected by the standardize output-paths and should work out-of-the-box.

Other than that works for me ✌🏻

inyourfoss committed:
alsa-scarlett-gui: Change icons path back to relative in desktop entry file

SebTM commented on pkgs/applications/audio/alsa-scarlett-gui/default.nix:
```diff
@@ -24,6 +24,11 @@ stdenv.mkDerivation rec {
+  postInstall = ''
+      substituteInPlace $out/share/applications/vu.b4.${pname}.desktop \
+        --replace "Exec=/bin/alsa-scarlett-gui" "Exec=$out/bin/${pname}" \
+        --replace "Icon=alsa-scarlett-gui" "Icon=${pname}"
```
This replaces "alsa-scarlett-gui" with "alsa-scarlett-gui"?

gnumber13 replied:
Yeah sorry those are remnants from testing that line will be removed. As its not needed.

inyourfoss committed:
alsa-scarlett-gui: light refactoring

SuperSandro2000 commented on pkgs/applications/audio/alsa-scarlett-gui/default.nix:
```diff
@@ -24,6 +24,10 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ pkg-config wrapGAppsHook4 ];
   buildInputs = [ gtk4 alsa-lib ];
+  postInstall = ''
+      substituteInPlace $out/share/applications/vu.b4.${pname}.desktop \
+        --replace "Exec=/bin/alsa-scarlett-gui" "Exec=$out/bin/${pname}"
```
Intendation is not matching 2 spaces which is written in our contributing guide and also we shouldn't substitute pname wherever possible

SebTM replied:
https://github.com/NixOS/nixpkgs/pull/246636

SebTM commented:
Hey @gnumber13, congratulations on your first contribution/pull-request to NixOS/nixpkgs 😃🎉

gnumber13 commented:
Hey @SebTM Thanks! And thank you for your support :)
